File: typescript/grimlock/utils/memory-security.ts
/**
 * Memory security utilities for Grimlock crypto module
 * 
 * Provides functions to securely erase sensitive data from memory.
 * Note: Complete memory erasure is difficult in JavaScript, but we
 * make best-effort attempts to zero out sensitive data.
 */

/**
 * Securely erase a Uint8Array by zeroing its contents
 * 
 * Note: In JavaScript, we cannot guarantee complete memory erasure
 * due to garbage collection and memory management. This function
 * makes a best-effort attempt to zero out the data.
 */
export function secureErase(data: Uint8Array): void {
  // Zero out all bytes
  data.fill(0);
  
  // In Node.js, we can try to force garbage collection hint
  // (though this is not guaranteed to work)
  if (typeof global !== 'undefined' && (global as any).gc) {
    // Only if --expose-gc flag is used
    try {
      (global as any).gc();
    } catch {
      // Ignore if gc is not available
    }
  }
}

/**
 * Securely erase multiple Uint8Arrays
 */
export function secureEraseMultiple(arrays: Uint8Array[]): void {
  for (const arr of arrays) {
    secureErase(arr);
  }
}

/**
 * Create a secure copy of data and erase the original
 * 
 * This is useful when you need to pass data but want to
 * ensure the original is zeroed.
 */
export function secureMove(data: Uint8Array): Uint8Array {
  const copy = new Uint8Array(data);
  secureErase(data);
  return copy;
}

File: python/grimlock/utils/memory_security.py
"""Memory security utilities for Grimlock crypto module.

Note: Python's memory management makes true secure erasure difficult,
but we provide best-effort clearing of sensitive data.
"""


def secure_erase(data: bytearray) -> None:
    """Securely erase sensitive data from memory (best-effort in Python).

    Note: Due to Python's memory management, we cannot guarantee complete
    erasure, but we clear the data to the best of our ability.

    Args:
        data: Bytearray to erase (must be mutable)
    """
    if isinstance(data, bytearray):
        data[:] = b"\x00" * len(data)
    elif isinstance(data, bytes):
        # For immutable bytes, we can't erase, but we try to clear references
        # The caller should use bytearray for mutable data
        pass


def secure_erase_multiple(*data: bytearray) -> None:
    """Securely erase multiple bytearrays.

    Args:
        *data: Variable number of bytearrays to erase
    """
    for d in data:
        secure_erase(d)
